### DISCUSSION TITLE
Record ForCtorParam MapFrom validation

### DISCUSSION THREAD

**Discussion by killergege** [LuckyPennySoftware/AutoMapper#4143] [NA]
### Source/destination types

```csharp
    private class Source
    {
        public string         Name     { get; init; }
        public SourceExpenses Expenses { get; init; }
    }

    private class SourceExpenses
    {
        public int Value1 { get; init; }
        public int Value2 { get; init; }
    }

    private record Target(string Name, int Expenses);
```

### Mapping configuration

```csharp
            CreateMap<Source, Target>(MemberList.Destination)
               .ForCtorParam(nameof(Source.Name), opts => opts.MapFrom(x => x.Name))
               .ForCtorParam(nameof(Source.Expenses), opts => opts.MapFrom((source, _) => source.Expenses.Value1 + source.Expenses.Value2));
```

### Expected behavior

Mapping works and `config.AssertConfigurationIsValid();` does not throw

### Actual behavior

Mapping works but `config.AssertConfigurationIsValid();` throws.
When fields have the same name but different type, it seems that the validator doesn't detect that the field mapping is overriden and tries to map the source type with the target type.

There is no issue when the `Target` object is a class, and the mapping is done with `ForMember` instead of `ForCtorParam`.
```
Unhandled exception. AutoMapper.AutoMapperConfigurationException: The following member on Target cannot be mapped: 
    Expenses 
Add a custom mapping expression, ignore, add a custom resolver, or modify the destination type Target.
Context:
    Mapping to member Expenses from Source to Target
```

### Steps to reproduce

Fiddle demo : https://dotnetfiddle.net/WoqHS4

**Comment by lbargaoanu** [comment]
The message is about the property `Expenses`, not about the constructor parameter. The record property is `init`, it matches, and it cannot be mapped. The easy way out is to rename the source property or the constructor parameter. A map from `SourceExpenses` to `int` would also work.
For the constructor parameter, the message is 
AutoMapper.AutoMapperConfigurationException: The following member on UserQuery+Target cannot be mapped:
  Void .ctor(System.String, Int32), parameter expenses

**Reply by lbargaoanu** [reply]
Record's weirdness aside, I guess if a property is mapped through the constructor, we can skip validating the property itself. You would get a runtime error if you bypass constructor mapping by passing an existing destination.
This new behaviour is in the MyGet build.

**Reply by jbogard** [reply]
I think that makes sense, but IIRC we've avoided that dual validation because we don't """know""" if a constructor parameter internally sets a property. But it's probably reasonable to make that assumption if the names match.

**Reply by lbargaoanu** [reply]
We kind of started [assuming](https://github.com/AutoMapper/AutoMapper/search?q=constructorparametermatches) :) It seems fair. AutoMapper relies on conventions and names are important.
